q2_mlab/db/maint.py

```diff
@@ -52,24 +56,46 @@ def add(engine: Engine, results: pd.DataFrame, parameters: dict,
     session.flush()
     params_id = params.id
 
-    # check if algorithm is valid
-    valid_algorithms = RegressionTask.algorithms
-    if algorithm not in valid_algorithms:
-        raise ValueError(f"Invalid choice '{algorithm}' for algorithm."
-                         f"Valid choices: {valid_algorithms}."
-                         )
+    # check if algorithm is valid, and of regression or classification
+    # and assign Score to the corresponding Regression or Classification table
+    valid_algorithms = set(RegressionTask.algorithms).union(
+        set(ClassificationTask.algorithms)
+    )
+    if algorithm in RegressionTask.algorithms:
+        Score = RegressionScore
+    elif algorithm in ClassificationTask.algorithms:
+        Score = ClassificationScore
+    else:
+        raise ValueError(
+            f"Invalid choice '{algorithm}' for algorithm."
+            f"Valid choices: {valid_algorithms}."
+        )
+
+    # check if uuid is in table, and skip it if so
+    # TODO once the code that pulls artifacts into the db and
+    # moves artifacts into an "inserted" directory is final, we
+    # may wish to change this to raise an error
+    query = session.query(Score).filter_by(
+        artifact_uuid=artifact_uuid
+    )
+    matching_artifact = query.first()
+    if matching_artifact is not None:
+        print(artifact_uuid + " already exists in table")
+        session.close()
+        return
```

Comment:
Printing an error message and returning `None` is undesirable for a couple reasons:
i) Returning `None` is the same as the default behavior of this function. If someone is using this function programmatically, and/or does not see/notice/etc. they receive no indication that their record was not inserted into the database. It would be more standard to raise and exception here.
ii) Printing to stdout is "messy" and limits client behavior. Say a user has an idea for what they want to be shown in the console, it is laborious for them to escape the print statements in your code. Alternatively, some applications depend on the fact that their application is the only program streaming to standard out. Python's `logging` module can be a viable alternative and let's user's decide the verbosity level that fits their needs, if used properly.

Reply:
Thanks! This makes a lot of sense. I'll add this as an optional check that raises an exception.